### PR TITLE
Refactor select functionality from `util` package to new `utilsaddr` package

### PR DESCRIPTION
**Pull request by jentfoo**
PR is best reviewed commit by commit.  There are two files which are being moved in this PR:
* `lib/utils/addr.go` - The `NetAddr` struct is referenced broadly across the project.  Having it in the `utils` package makes it difficult to avoid potential cyclic import issues with the `util` package.  This came up trying to adapt the feedback in this PR: https://github.com/gravitational/teleport/pull/30184#discussion_r1287482688

In this change it has moved to `lib/utils/utilsaddr/addr.go`.  The new `utilsaddr` package is brief when referenced and avoids the cyclic import risks.

* `lib/utils/net.go` - I looked for other files which may be a good candidate to move to this package.  This one was a good fit, although some functionality was split to the `addr.go` file.  The bulk of the functionality was renamed into a more precisely named `utilsaddr/context.go`

PR for `e` changes to make temporary alias no longer necessary: https://github.com/gravitational/teleport.e/pull/1963

I have added the backport labels since the `NetAddr` is so broadly used, and I am concerned this change will cause friction when backporting.

### CODE REVIEW

File: lib/utils/addr/addr.go [hunk 1]
```diff
@@ -355,3 +355,15 @@ func ReplaceUnspecifiedHost(addr *NetAddr, defaultPort int) string {
 	port := addr.Port(defaultPort)
 	return net.JoinHostPort("localhost", strconv.Itoa(port))
 }
+
+// ClientIPFromConn extracts host from provided remote address.
+func ClientIPFromConn(conn net.Conn) (string, error) {
```
Comment (codingllama) [inline lib/utils/addr/addr.go:360]: Move func to its own file? It seems disjointed from the rest.

File: lib/utils/addr/addr.go
Comment (codingllama) [inline lib/utils/addr/addr.go:1]: In the original file/package:

```go
// TODO(jentfoo): Remove once e/ is updated to use utils/addr.
type NetAddr = addr.NetAddr
type ParseAddr = addr.ParseAddr
// (so on, whatever is needed to not break)
```

**Review by codingllama — COMMENTED** [review]
LGTM, minus the potential build breakages.

+1 for backporting, so other backports are not impacted.

+1 for smaller, more focused utility packages.

File: lib/utils/addr/addr.go (hunk 1 above)
Comment (jentfoo) [reply lib/utils/addr/addr.go:360]: It seemed related in the sense that it's IP parsing.  I can put it in a separate file, but I felt it was related enough to combine.  Let me know your preference

**Review by codingllama — APPROVED** [review]
Thanks for taking the time to avoid breakages.

**Comment by jentfoo** [comment]
I updated the package from `addr` to `utilsaddr`.  `netloc` and `netinfo` were also considered, but after discussing with @codingllama we decided to just go with the more descriptive name.

This was needed because there is a lot of `addr` variables that are complicating the refactor.

**Review by greedy52 — APPROVED** [review]
LGTM.

`lib/utils/utilsaddr` still depends on `api/utils` though. what do you guys think if we move `lib/utils/utilsaddr` and `api/utils/addr.go` to `api/utils/utilsaddr`. I am fine with current change too.

**Review by codingllama — APPROVED** [review]

**Comment by jentfoo** [comment]
> lib/utils/utilsaddr still depends on api/utils though

I will need a follow up PR to remove `addr_alias`.  Just trying to get the build working (which it's not, I still need to figure out the flakey tests and integration tests, the failure is non-obvious to me)

**Comment by jentfoo** [comment]
Closing, this PR has significantly rotted again.  It does not seem the effort of continual rebasing is worth it in this case.  However I am happy to re-open it if anyone is interested in paring on the Flakey Test failures.

**Pull request closed without merging**